src/ListT/Weighted.hs

```diff
@@ -0,0 +1,122 @@
+{-# LANGUAGE FlexibleInstances #-}
+-- {-# LANGUAGE InstanceSigs #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+-- | A re-implementation of @weighted-search@
+-- as a monad transformer.
+module ListT.Weighted
+  ( Weight (..),
+    WeightedT,
+    Weighted,
+    weight,
+    (<|>),
+    toList,
+  )
+where
+
+import Control.Applicative (Alternative (..))
+import Control.Monad (MonadPlus (..), ap)
+import Control.Monad.IO.Class (MonadIO (..))
+import Control.Monad.Reader.Class (MonadReader (..))
+import Control.Monad.State.Class (MonadState (..))
+import Control.Monad.Trans (MonadTrans (..))
+import Control.Monad.WeightedSearch (Weight (..))
+import Data.Foldable (fold, toList)
+import Data.Functor.Identity (Identity)
+
+-- | Weighted nondeterminstic computations over the weight @w@.
+data WeightedT w m a
+  = Fail
+  | Yield a (WeightedT w m a)
+  | Weight w (WeightedT w m a)
+  | Action (m (WeightedT w m a))
+
+type Weighted w = WeightedT w Identity
+
+weight :: w -> WeightedT w m a -> WeightedT w m a
+weight = Weight
+
+mapWeightedT :: (Functor n) => (forall a. m a -> n a) -> WeightedT w m b -> WeightedT w n b
+mapWeightedT _ Fail = Fail
+mapWeightedT f (Yield x m) = Yield x (mapWeightedT f m)
+mapWeightedT f (Weight w m) = Weight w (mapWeightedT f m)
+mapWeightedT f (Action a) = Action (mapWeightedT f <$> f a)
+
+instance (Functor m) => Functor (WeightedT w m) where
+  fmap _ Fail = Fail
+  fmap f (Yield x w) = Yield (f x) (fmap f w)
+  fmap f (Weight a w) = Weight a (fmap f w)
+  fmap f (Action w) = Action (fmap f <$> w)
+
+instance (Weight w, Applicative m) => Applicative (WeightedT w m) where
+  pure = return
+  (<*>) = ap
+
+instance (Weight w, Applicative m) => Monad (WeightedT w m) where
+  return x = Yield x Fail
+  Fail >>= _ = Fail
+  Yield x m >>= f = f x `mplus` (m >>= f)
+  Weight w m >>= f = Weight w (m >>= f)
+  Action w >>= f = Action ((>>= f) <$> w)
+
+instance (Weight w, Applicative m) => Alternative (WeightedT w m) where
+  empty = mzero
+  (<|>) = mplus
+
+instance (Weight w, Applicative m) => MonadPlus (WeightedT w m) where
+  mzero = Fail
+  Fail `mplus` m = m
+  Yield x m `mplus` n = Yield x (m `mplus` n)
+  Weight w m `mplus` Action a = Action (mplus (Weight w m) <$> a)
+  Action w `mplus` m = Action (mplus <$> w <*> pure m)
+  Weight w m `mplus` Fail = Weight w m
+  Weight w m `mplus` Yield x n = Yield x (Weight w m `mplus` n)
+  Weight w m `mplus` Weight w' n =
+    case compare w w' of
+      LT -> Weight w (m `mplus` Weight (difference w' w) n)
+      EQ -> Weight w (m `mplus` n)
+      GT -> Weight w' (Weight (difference w w') m `mplus` n)
```

Comment:
I feel there's black magic on this instance, why don't we have a ``m `mplus` Fail = m`` clause?

Reply:
That's the one I copied from the original package, so I don't have full insight about it yet.

I'm pretty sure that line you mention could be added; but I'm a bit concerned about potential loss of laziness in the case in which we `Yield` a nested amount of elements, since in that case the current definition of `mplus` will never look at the other side until we have consumed all those elements.

Reply:
That's a very good concern!

Reply:
But in the case ``Action m `mplus` Yield x n`` we're not yielding before executing the action, even though we could.

Reply:
I think that this change the semantic.  ```Action Nothing `mplus` Yield x n``` currently outputs `Action Nothing`, with this change it would outputs `Yield x (Action Nothing)` (assuming `n` is not a `Yield` itself).

Reply:
Mmmm... after thinking a bit more, I think that indeed `mplus (Action m) thing` should always first execute the `action` before `mplus`ing with `thing`. So I'm leaving it as it was.

Reply:
> I think that this change the semantic. Action Nothing `mplus` Yield x n currently outputs Action Nothing, with this change it would outputs Yield x (Action Nothing) (assuming n is not a Yield itself).

I don't think so... say: ``Action Nothing `mplus` Yield x n :: WeightedT Maybe ()`, the denotation is at `Maybe [()]`,
```
denot (Action Nothing) == Nothing
denot (Yield x (Action Nothing)) == (x :) <$> denot (Action Nothing) == Nothing
```

What it changes is the order in which effects get executed